src/hotspot/share/utilities/count_trailing_zeros.hpp

```diff
@@ -59,13 +65,23 @@ inline unsigned count_trailing_zeros(uintx x) {
 #pragma intrinsic(_BitScanForward)
 #endif
 
-inline unsigned count_trailing_zeros(uintx x) {
-  assert(x != 0, "precondition");
+inline unsigned count_trailing_zeros_32(uint32_t x) {
+  unsigned long index;
+  _BitScanForward(&index, x);
+  return index;
+}
+
+inline unsigned count_trailing_zeros_64(uint64_t x) {
   unsigned long index;
 #ifdef _LP64
   _BitScanForward64(&index, x);
 #else
-  _BitScanForward(&index, x);
+  if (_BitScanForward(&index, (uint32_t)x) == 0) {
+    // no bit found? If so, try the upper dword. Otherwise index already contains the result
+    _BitScanForward(&index, x >> 32);
```

Comment:
If the first BSF needs a cast then this one probably does too.  The shift result is still uint64_t.